### PR TITLE
84304 - rustdoc: shrink Item::Attributes

### PR DESCRIPTION
**Pull request by tdelabro**
Helps with https://github.com/rust-lang/rust/issues/84304

### CODE REVIEW

File: src/librustdoc/clean/types.rs [hunk 1]
```diff
@@ -648,12 +647,38 @@ impl<'a> Iterator for ListAttributesIter<'a> {
...
 impl AttributesExt for [ast::Attribute] {
     fn lists(&self, name: Symbol) -> ListAttributesIter<'_> {
         ListAttributesIter { attrs: self.iter(), current_list: Vec::new().into_iter(), name }
     }
+    
+    fn span(&self) -> Option<rustc_span::Span> {
+        self.iter()
+            .filter(|attr| attr.doc_str().is_some())
+            .map(|attr| attr.span)
+            .next()
```
Comment (jyn514) [inline src/librustdoc/clean/types.rs:667]:
```suggestion
            .find(|attr| attr.doc_str().is_some())
            .map(|attr| attr.span)
```

File: src/librustdoc/clean/types.rs [hunk 2]
```diff
@@ -648,12 +647,38 @@ impl<'a> Iterator for ListAttributesIter<'a> {
...
+    
+    fn other_attrs(&self) -> Vec<ast::Attribute>;
 }
 
 impl AttributesExt for [ast::Attribute] {
     fn lists(&self, name: Symbol) -> ListAttributesIter<'_> {
         ListAttributesIter { attrs: self.iter(), current_list: Vec::new().into_iter(), name }
     }
+    
+    fn span(&self) -> Option<rustc_span::Span> {
```
Comment (jyn514) [inline src/librustdoc/clean/types.rs:663]: Wow, I'm just now realizing this is the span of the *doc string*, not of any other part of the attribute. Could you add that as a comment?
```suggestion
    /// Return the span of the first doc-comment, if it exists.
    fn span(&self) -> Option<rustc_span::Span> {
```

File: src/librustdoc/clean/types.rs [hunk 3]
```diff
@@ -648,12 +647,38 @@ impl<'a> Iterator for ListAttributesIter<'a> {
...
+        self.iter()
+            .filter(|attr| attr.doc_str().is_some())
+            .map(|attr| attr.span)
+            .next()
+    }
+
+    fn inner_docs(&self) -> bool {
```
Comment (jyn514) [inline src/librustdoc/clean/types.rs:670]:
```suggestion
    /// Returns whether the first doc-comment is an inner attribute.
    ///
    //// If there are no doc-comments, return true.
    /// FIXME(#78591): Support both inner and outer attributes on the same item.
    fn inner_docs(&self) -> bool {
```

File: src/librustdoc/clean/types.rs [hunk 4]
```diff
@@ -648,12 +647,38 @@ impl<'a> Iterator for ListAttributesIter<'a> {
...
+        .find(|a| a.doc_str().is_some())
+        .map_or(true, |a| a.style == AttrStyle::Inner)
+    }
+
+    fn other_attrs(&self) -> Vec<ast::Attribute> {
+        self.iter().filter_map(|attr| match attr.doc_str() {
+            Some(_) => None,
+            None => Some(attr.clone()),
+        }).collect()
+    }
```
Comment (jyn514) [inline src/librustdoc/clean/types.rs:681]: Hmm, you added this function, but didn't remove the `other_attrs` field - was that intentional?

File: src/librustdoc/clean/types.rs [hunk 5]
```diff
@@ -648,12 +647,38 @@ impl<'a> Iterator for ListAttributesIter<'a> {
...
+        self.iter()
+        .find(|a| a.doc_str().is_some())
+        .map_or(true, |a| a.style == AttrStyle::Inner)
+    }
+
+    fn other_attrs(&self) -> Vec<ast::Attribute> {
+        self.iter().filter_map(|attr| match attr.doc_str() {
+            Some(_) => None,
+            None => Some(attr.clone()),
+        }).collect()
```
Comment (jyn514) [inline src/librustdoc/clean/types.rs:680]:
```suggestion
    fn other_attrs(&self) -> impl Iterator<Item = &ast::Attribute> {
        self.iter().filter(|attr| attr.doc_str().is_none())
```

Comment (jyn514) [reply src/librustdoc/clean/types.rs:680]: Hmm, I realized while reviewing that `attr.doc_str().is_some()` and `attr.is_doc_comment()` behave differently :/ I'll look into fixing that.

File: src/librustdoc/clean/types.rs [hunk 6]
```diff
@@ -986,17 +1005,10 @@ impl Attributes {
             }
         }
 
-        let inner_docs = attrs
-            .iter()
-            .find(|a| a.doc_str().is_some())
-            .map_or(true, |a| a.style == AttrStyle::Inner);
-
         Attributes {
             doc_strings,
```
Comment (jyn514) [inline src/librustdoc/clean/types.rs:1009]: Would it possible to also calculate this on demand? I think `extract_include` reads from the filesystem, so you probably want to add a cache - I'm fine with putting that off for a later PR.

File: src/librustdoc/clean/types.rs [hunk 7]
```diff
@@ -1090,6 +1100,18 @@ impl AttributesExt for Attributes {
...
+        todo!("delete");
+    }
+    
+    fn inner_docs(&self) -> bool {
+        todo!("delete");
+    }
+    
+    fn other_attrs(&self) -> Vec<ast::Attribute> {
+        self.other_attrs.clone()
+    }
```
Comment (jyn514) [inline src/librustdoc/clean/types.rs:1114]: Rather than add functions that aren't supported, do you think it makes sense *not* to implement AttributesExt for Attributes, and add `lists` and `other_attrs` as inherent functions? That would give a compile error if you tried to call an unsupported functoin.

File: src/librustdoc/html/render/cache.rs [hunk 8]
```diff
@@ -50,7 +52,7 @@ crate fn extern_location(
 
     // Failing that, see if there's an attribute specifying where to find this
     // external crate
-    e.attrs
+    ast_attrs
```
Comment (jyn514) [inline src/librustdoc/html/render/cache.rs:55]: It looks like this was the only place the attributes for ExternalCrate was used, so you could turn it into a parameter instead - that's a great change, but it makes it a little harder to see what changes are related to removing files from Attributes itself. Do you mind splitting it into a separate commit?

File: src/librustdoc/passes/collect_intra_doc_links.rs [hunk 9]
```diff
@@ -855,7 +856,10 @@ impl<'a, 'tcx> DocFolder for LinkCollector<'a, 'tcx> {
             }
         });
 
-        if item.is_mod() && item.attrs.inner_docs {
+        let ast_attrs: &[ast::Attribute] = &self.cx.tcx.get_attrs(item.def_id);
+        let inner_docs = ast_attrs.inner_docs();
```
Comment (jyn514) [inline src/librustdoc/passes/collect_intra_doc_links.rs:860]:
```suggestion
        let inner_docs = self.cx.tcx.get_attrs(item.def_id).inner_docs();
```

File: src/librustdoc/passes/collect_intra_doc_links.rs [hunk 10]
```diff
@@ -1052,6 +1056,10 @@ impl LinkCollector<'_, '_> {
             };
         let mut path_str = &*path_str;
 
+
+        let ast_attrs: &[ast::Attribute] = &self.cx.tcx.get_attrs(item.def_id);
+        let inner_docs = ast_attrs.inner_docs();
```
Comment (jyn514) [inline src/librustdoc/passes/collect_intra_doc_links.rs:1061]:
```suggestion
        let inner_docs = self.cx.tcx.get_attrs(item.def_id).inner_docs();
```

**Review by jyn514 — COMMENTED** [review]
This is a great change :) but I think the title is not quite accurate? It should say something like "Shrink Item::Attributes", Attributes still calculates some info ahead of time.

To fix the CI error you can run `x.py fmt`.

File: src/librustdoc/passes/collect_intra_doc_links.rs [hunk 11]
```diff
@@ -856,8 +856,10 @@ impl<'a, 'tcx> DocFolder for LinkCollector<'a, 'tcx> {
             }
         });
 
-        let ast_attrs: &[ast::Attribute] = &self.cx.tcx.get_attrs(item.def_id);
-        let inner_docs = ast_attrs.inner_docs();
+        let inner_docs = match self_id {
+            Some(did) => self.cx.tcx.get_attrs(did).inner_docs(),
+            None => false,
+        };
```
Comment (jyn514) [inline src/librustdoc/passes/collect_intra_doc_links.rs:862]: This is wrong. It will break if this is documenting a function and not a type. You should check item.is_fake instead.

**Comment by jyn514** [comment]
@tdelabro do you mind using rebasing instead of merging? There are instructions in https://rustc-dev-guide.rust-lang.org/git.html#conflicts.

File: src/librustdoc/passes/collect_intra_doc_links.rs (hunk 11 above)
Comment (tdelabro) [reply src/librustdoc/passes/collect_intra_doc_links.rs:862]:
```rust
let self_id = if item.is_fake() {
            None
        // Checking if the item is a field in an enum variant
        }
```
`self_id` is already the result of testing `item.is_fake()`. How is using it different than checking `item.is_fake()` again ?

Comment (jyn514) [reply src/librustdoc/passes/collect_intra_doc_links.rs:862]: `tcx.parent` can return `None`. And this is not really the check you want anyway, the self ID is irrelevant when what you actually care about is whether the item is fake.

Comment (tdelabro) [reply src/librustdoc/passes/collect_intra_doc_links.rs]: Ok. It's done

**Comment by tdelabro** [comment]
> @tdelabro do you mind using rebasing instead of merging? There are instructions in https://rustc-dev-guide.rust-lang.org/git.html#conflicts.

Sorry about this. I fixed it with a push force.

File: src/librustdoc/clean/types.rs (hunk 5 above)
Comment (jyn514) [reply src/librustdoc/clean/types.rs]: I noticed you change this to use `.filter()` but didn't change the return type - was that intentional? Did you run into trouble?

Comment (jyn514) [reply src/librustdoc/clean/types.rs]: The reason I suggested removing `.collect()` is it avoids an allocation for the common case where you just want to check each attribute one at a time.

File: src/librustdoc/formats/cache.rs [hunk 12]
```diff
@@ -164,10 +164,12 @@ impl Cache {
...
             let name = e.name(tcx);
             let extern_url = extern_html_root_urls.get(&*name.as_str()).map(|u| &**u);
-            self.extern_locations
-                .insert(n, (name, src_root, extern_location(e, extern_url, &dst, tcx)));
-
             let did = DefId { krate: n, index: CRATE_DEF_INDEX };
+            self.extern_locations.insert(
+                n,
+                (name, src_root, extern_location(e, extern_url, tcx.get_attrs(did), &dst, tcx)),
+            );
```
Comment (jyn514) [inline src/librustdoc/formats/cache.rs:171]: Doesn't have to be here, but it would be nice to store the CrateNum here instead of the name and attributes. Both of those are available from the tcx.

Actually all of this looks like it could be calculated on-demand with only the CrateNum, tcx, and extern_html_root_urls.

**Comment by tdelabro** [comment]
At this point `Attributes` is only two fields large:
```rust
#[derive(Clone, Debug, Default)]
crate struct Attributes {
    crate doc_strings: Vec<DocFragment>,
    crate other_attrs: Vec<ast::Attribute>,
}
```
Next I will try to store `other_attrs` in `Item` so they will be available at any moment to build `doc_strings`, which is based on the combination of the `Item` attributes and these `other_attrs`. Once this will be done, I think we will be able to remove `Attributes` form the codebase. 

However, I don't know what we should do with all the functions defined in `impl Attributes` (eg: `extract_include` or `has_doc_flags`). Probably move them to `impl Item`.
What do you think ?

**Review by jyn514 — COMMENTED** [review]
> Next I will try to store other_attrs in Item so they will be available at any moment to build doc_strings, which is based on the combination of the Item attributes and these other_attrs. Once this will be done, I think we will be able to remove Attributes form the codebase.

I think we should save this for a follow-up PR, it will be easier to catch regressions.

> However, I don't know what we should do with all the functions defined in impl Attributes (eg: extract_include or has_doc_flags). Probably move them to impl Item.

I would move them to AttributesExt where possible.

File: src/librustdoc/formats/cache.rs (hunk 12 above)
Comment (tdelabro) [reply src/librustdoc/formats/cache.rs:171]: I'm not sure I understood correctly:
Yo would like to replace `crate external_paths: FxHashMap<DefId, (Vec<String>, ItemType)>` by `crate extern_html_urls: FxHashMap<DefId, &BTreeMap<String, String>>`, get `name` and `attributes` from `tcx`, combine them with `cache.extern_html_root_url[CrateNum]` and then compute `extern_locations` with all this ?

Comment (jyn514) [reply src/librustdoc/formats/cache.rs:171]: I would like to get rid of `extern_locations` altogether, and calculate `name`, `src_root`, and `external_location` on demand. `external_paths` can stay the same.

I'll open an issue for it, it doesn't belong here.

File: src/librustdoc/clean/types.rs [hunk 13]
```diff
@@ -940,19 +1009,14 @@ impl Attributes {
...
-                            // Extracted #[doc(cfg(...))]
-                            match Cfg::parse(cfg_mi) {
-                                Ok(new_cfg) => cfg &= new_cfg,
-                                Err(e) => diagnostic.span_err(e.span, e.msg),
-                            }
+                        if let Some(_) = Attributes::extract_cfg(&mi) {
+                            // There used to be cfg related there. It had been mooved.
+                            // The second member used to only execute if this one was false (`else if`).
+                            // Is it the flow we want to keep, or can we get rid of this first condition without repercusions ?
                         } else if let Some((filename, contents)) = Attributes::extract_include(&mi)
```
[the remark this answers is not in the document]
Comment (tdelabro) [reply src/librustdoc/clean/types.rs]: Done

File: src/librustdoc/clean/types.rs (hunk 6 above)
Comment (tdelabro) [reply src/librustdoc/clean/types.rs]: It had been moved as an `AttributeExt` method

File: src/librustdoc/html/render/cache.rs (hunk 8 above)
Comment (tdelabro) [reply src/librustdoc/html/render/cache.rs:55]: What should I do here ? Revert the changes on `ExternalCrate` and put them in a separate MR ?

Also I didn't really get the "turn it into a parameter instead" part. Can you give me a little bit more explanation ?

File: src/librustdoc/clean/types.rs (hunk 5 above)
Comment (tdelabro) [reply src/librustdoc/clean/types.rs]: Ho ok. That's why !

I didn't notice the change on the return type, so yes I got some trouble and had to clone and collect the attributes. 
I will try your way instead !

File: src/librustdoc/clean/types.rs (hunk 4 above)
Comment (tdelabro) [reply src/librustdoc/clean/types.rs]: This implementation was required to remove `cfg` from `Attributes` but to be able to entirely remove the `other_attrs` field from `Attributes` we will have to move it in `Item`. It will be done in a future MR.

File: src/librustdoc/clean/types.rs (hunk 5 above)
Comment (tdelabro) [reply src/librustdoc/clean/types.rs]: I tried, didn't go well ;)

File: src/librustdoc/formats/cache.rs (hunk 12 above)
Comment (jyn514) [reply src/librustdoc/formats/cache.rs:171]: Oipened https://github.com/rust-lang/rust/issues/84588.

File: src/librustdoc/html/render/cache.rs (hunk 8 above)
Comment (jyn514) [reply src/librustdoc/html/render/cache.rs:55]: > Should I revert the changes on ExternalCrate and put them in a separate MR ?

Nah, don't worry about it, this is fine :) It's a good change.

> Also I didn't really get the "turn it into a parameter instead" part. Can you give me a little bit more explanation ?

This is something you're already doing: instead of storing the attributes on ExternalCrate, you pass them in to `extern_location`, which is the only place that uses them.

File: src/librustdoc/html/render/print_item.rs [hunk 14]
```diff
@@ -284,16 +286,14 @@ fn item_module(w: &mut Buffer, cx: &Context<'_>, item: &clean::Item, items: &[cl
...
-                        None,
-                    ));
+                    let ast_attrs = cx.tcx().get_attrs(import_def_id);
+                    let import_attrs = Box::new(clean::Attributes::from_ast(ast_attrs, None));
 
                     // Just need an item with the correct def_id and attrs
                     let import_item = clean::Item {
                         def_id: import_def_id,
                         attrs: import_attrs,
+                        cfg: ast_attrs.cfg(cx.tcx().sess().diagnostic()),
```
Comment (jyn514) [inline src/librustdoc/html/render/print_item.rs:296]: The function call is unnecessary.
```suggestion
                        cfg: ast_attrs.cfg(cx.tcx().sess.diagnostic()),
```

File: src/librustdoc/passes/collect_intra_doc_links.rs [hunk 15]
```diff
@@ -1050,6 +1054,8 @@ impl LinkCollector<'_, '_> {
             };
         let mut path_str = &*path_str;
 
+        let inner_docs = self.cx.tcx.get_attrs(item.def_id).inner_docs();
```
Comment (jyn514) [inline src/librustdoc/passes/collect_intra_doc_links.rs:1057]: Doesn't this also need to check for `is_fake()`? Maybe it would make sense to add an `Item::inner_docs` function to do the check automatically?

File: src/librustdoc/clean/inline.rs [hunk 16]
```diff
@@ -122,10 +122,16 @@ crate fn try_inline(
 
     let target_attrs = load_attrs(cx, did);
     let attrs = box merge_attrs(cx, Some(parent_module), target_attrs, attrs_clone);
+    let cfg = if let Some(new_attrs) = attrs_clone {
+        let mut both = new_attrs.to_vec();
+        both.extend_from_slice(target_attrs);
+        both.cfg(cx.sess().diagnostic())
+    } else {
+        target_attrs.cfg(cx.sess().diagnostic())
+    };
```
[the remark this answers is not in the document]
Comment (tdelabro) [reply src/librustdoc/clean/inline.rs:131]: `attrs` is of type `clean::Attributes`. The whole point of this series of changes was to take `cfg` out of `clean::Attributes`, so we cannot do `attrs.cfg()` anymore. It's now either stored in `Item` or computed on the run from `[ast::attributes]`. 
So here we want to compute them from attributes in order to store them in the `Item` we are building and we are facing the case where `clean::Attributes` is build from two sets of `[ast::attribute]` (`attrs` and `additional_attrs` as they are named in `clean::Attributes::from_ast`).

In `from_ast` the order was relevant, `additional_attrs` must be placed before `attrs` in the resulting `Attributes`, but I realize now that if we just compute `cfg` the order may not matter anymore. But I'm not entirely sure because it will influence the order in which the different `Cfg` are pushed into the `Cfg::All` vector. Does it matter latter ?

File: src/librustdoc/clean/inline.rs [hunk 17]
```diff
@@ -415,8 +420,16 @@ crate fn build_impl(
...
+    let old_attrs = load_attrs(cx, did);
+    let merged_attrs = box merge_attrs(cx, parent_module.into(), old_attrs, attrs);
     debug!("merged_attrs={:?}", attrs);
+    let cfg = if let Some(new_attrs) = attrs {
+        let mut both = new_attrs.to_vec();
+        both.extend_from_slice(old_attrs);
+        both.cfg(cx.sess().diagnostic())
+    } else {
+        old_attrs.cfg(cx.sess().diagnostic())
+    };
```
[the remark this answers is not in the document]
Comment (tdelabro) [reply src/librustdoc/clean/inline.rs:432]: And I would totally agree to factorize this redundancy. Should I create a dedicated function in inline.rs ? Similar to the `merge_attrs` one, but for cfg.

File: src/librustdoc/clean/types.rs [hunk 18]
```diff
@@ -663,12 +670,76 @@ impl<'a> Iterator for ListAttributesIter<'a> {
...
+                                Ok(new_cfg) => cfg &= new_cfg,
+                                Err(e) => diagnostic.span_err(e.span, e.msg),
+                            }
+                        }
+                    }
+                }
+            }
+        };
+
+        self.iter().for_each(collect_cfg);
```
[the remark this answers is not in the document]
Comment (tdelabro) [reply src/librustdoc/clean/types.rs:724]: I stuck to close to what this code was when it was inside `from_ast`. I will refacto this the way you suggest. Would be a lot nicer.

File: src/librustdoc/passes/collect_intra_doc_links.rs (hunk 15 above)
Comment (tdelabro) [reply src/librustdoc/passes/collect_intra_doc_links.rs]: For some reason I don't understand it wasn't causing any crash, but it could have. I did as you sad and it should be safe now 👍

File: src/librustdoc/html/render/print_item.rs (hunk 14 above)
Comment (tdelabro) [reply src/librustdoc/html/render/print_item.rs]: It's done. What are is the difference between calling the function and the field ? When should I use one over an other ?

File: src/librustdoc/clean/types.rs (hunk 18 above)
[the remark this answers is not in the document]
Comment (tdelabro) [reply src/librustdoc/clean/types.rs]: Done

File: src/librustdoc/clean/inline.rs [hunk 19]
```diff
@@ -444,7 +458,6 @@ fn build_module(
     visited: &mut FxHashSet<DefId>,
 ) -> clean::Module {
     let mut items = Vec::new();
-
```
[the remark this answers is not in the document]
Comment (tdelabro) [reply src/librustdoc/clean/inline.rs]: It's back as it was :)

**Comment by GuillaumeGomez** [comment]
Nice improvement on both the RSS usage and on the number of instructions (max -1.0% in both cases).

File: src/librustdoc/html/render/print_item.rs (hunk 14 above)
Comment (jyn514) [reply src/librustdoc/html/render/print_item.rs]: There's no difference. You should always use the field unless you're in a generic context where you don't know if you have a TyCtxt or not.

File: src/librustdoc/clean/inline.rs (hunk 16 above)
[the remark this answers is not in the document]
Comment (jyn514) [reply src/librustdoc/clean/inline.rs:131]: > In from_ast the order was relevant, additional_attrs must be placed before attrs in the resulting Attributes, but I realize now that if we just compute cfg the order may not matter anymore. 

Correct, the order doesn't matter - the only things you can do with `Cfg` is logical AND and OR.

> The whole point of this series of changes was to take cfg out of clean::Attributes, so we cannot do attrs.cfg() anymore. It's now either stored in Item or computed on the run from [ast::attributes].

Ok I see - and adding this function to clean::Attributes makes it confusing (because now there are 3 different places you could look up the cfg, but you should use `Item::cfg` wherever possible. Maybe it makes sense to return the cfg of both attributes from `merge_attrs` directly? Then you get rid of the duplication and it's also more clear why this is separate from everything else.

File: src/librustdoc/html/render/print_item.rs (hunk 14 above)
Comment (tdelabro) [reply src/librustdoc/html/render/print_item.rs]: Ok thanks

File: src/librustdoc/clean/inline.rs (hunk 16 above)
[the remark this answers is not in the document]
Comment (tdelabro) [reply src/librustdoc/clean/inline.rs:131]: > Maybe it makes sense to return the cfg of both attributes from merge_attrs directly? Then you get rid of the duplication and it's also more clear why this is separate from everything else.

Yes it makes sense. I'm on it

[the remark this answers is not in the document]
Comment (tdelabro) [reply src/librustdoc/clean/inline.rs]: It's done 👍

File: src/librustdoc/clean/inline.rs [hunk 20]
```diff
@@ -295,21 +294,24 @@ fn merge_attrs(
...
-            both.clean(cx)
-        }
+        let mut both = inner.to_vec();
+        both.extend_from_slice(old_attrs);
+        (
+            if let Some(new_id) = parent_module {
+                Attributes::from_ast(old_attrs, Some((inner, new_id)))
+            } else {
+                both.clean(cx)
+            },
```
Comment (jyn514) [inline src/librustdoc/clean/inline.rs:310]: It was like this before, but I think it would be more clear to use `from_ast` consistently:
```suggestion
            if let Some(new_id) = parent_module {
                Attributes::from_ast(old_attrs, Some((inner, new_id)))
            } else {
                Attributes::from_ast(both, None)
            },
```

**Review by jyn514 — COMMENTED** [review]
r=me with the nit addressed :)

File: src/librustdoc/clean/inline.rs (hunk 20 above)
Comment (tdelabro) [reply src/librustdoc/clean/inline.rs]: Indeed it's better this way

**Comment by tdelabro** [comment]
> Oh, and if you could squash the commits a little that would be nice too, but I won't block on it.

Here we are. I just kept two of them
